pyccl/baryons.py

```diff
@@ -0,0 +1,137 @@
+    """
+
+    def apply_baryons(self, cosmo, pk2d, in_place=False):
+        """Apply baryonic effects to a given power spectrum.
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): Cosmological parameters.
+            pk2d (:class:`~pyccl.pk2d.Pk2D`): power spectrum.
+            in_place (:obj:`bool`): if True, the `pk2d` itself is modified,
+                instead of returning a new `ccl.Pk2D` object.
+        """
+        raise NotImplementedError("Do not use base `BaryonicEffects` class.")
+
+    @unlock_instance(mutate=True, argv=2)
+    def _new_pk(self, cosmo, pk2d, a_arr, lk_arr, pk_arr, logp, in_place):
+        # Returns a new Pk2D, either in place or from scratch.
+        if in_place:
+            lib.f2d_t_free(pk2d.psp)
+            status = 0
+            pk2d.psp, status = lib.set_pk2d_new_from_arrays(
+                lk_arr, a_arr, pk_arr.flatten(),
+                int(pk2d.extrap_order_lok),
+                int(pk2d.extrap_order_hik),
+                int(logp), status)
+            check(status, cosmo)
+            return None
+        else:
+            new = Pk2D(a_arr=a_arr, lk_arr=lk_arr, pk_arr=pk_arr,
+                       is_logp=logp,
+                       extrap_order_lok=pk2d.extrap_order_lok,
+                       extrap_order_hik=pk2d.extrap_order_hik)
+            return new
+
+
+class BaryonicEffectsBCM(BaryonicEffects):
+    """The BCM model boost factor for baryons.
+
+    .. note:: BCM stands for the "baryonic correction model" of Schneider &
+              Teyssier (2015; https://arxiv.org/abs/1510.06034). See the
+              `DESC Note <https://github.com/LSSTDESC/CCL/blob/master/doc\
+/0000-ccl_note/main.pdf>`_
+              for details.
+
+    .. note:: The boost factor is applied multiplicatively so that
+              :math:`P_{\\rm corrected}(k, a) = P(k, a)\\, f_{\\rm bcm}(k, a)`.
+
+    Args:
+        log10Mc (:obj:`float`): logarithmic mass scale of hot
+            gas suppression. Defaults to 14.08.
+        eta_b (:obj:`float`): ratio of escape to ejection radii (see
+            Teyssier et al. 2015). Defaults to 0.5.
+        k_s (:obj:`float`): Characteristic scale (wavenumber) of
+            the stellar component. Defaults to 55.0.
+    """
+    __repr_attrs__ = ("log10Mc", "eta_b", "k_s")
+
+    def __init__(self, log10Mc=14.08, eta_b=0.5, k_s=55.0):
+        self.log10Mc = log10Mc
+        self.eta_b = eta_b
+        self.k_s = k_s
+        self.apply_baryons = self._apply_boost
+
+    def boost_factor(self, cosmo, k, a):
+        """The BCM model boost factor for baryons.
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): Cosmological parameters.
+            k (float or array_like): Wavenumber; Mpc^-1.
+            a (float): Scale factor.
+
+        Returns:
+            float or array_like: Correction factor to apply to
+                the power spectrum.
+        """
+        k_use = np.atleast_1d(k)
```

Comment:
If we don't use the power of `numpy` here we'll end up using for-loops in Python later, which is a no-go zone, particularly for "fast" code we ported from C. We have to make both input arrays [broadcastable](https://numpy.org/doc/stable/user/basics.broadcasting.html), so replace this with:
```python
        a, k = map(np.atleast_1d, [a, k])
        a, k = a[:, None], k[None, :]
```
and now both inputs can be floats or arrays. Also, to make reading the code easier, avoid using `a_use`, `k_use` etc. since we don't operate on them in-place.

Reply:
Note that we need to keep track of the original shape of k and a to return in the right shape. So partially done (still use `a_use` and `k_use` to keep track of the original `k` and `a`)